client/asset/eth/eth.go

```diff
@@ -3691,20 +3753,125 @@ func (w *assetWallet) checkFindRedemptions() {
 	}
 }
 
+// sumPendingTxs sums the expected incoming and outgoing values in unconfirmed
+// transactions stored in pendingTxs. Not used if the node is a txPoolFetcher.
+func (w *assetWallet) sumPendingTxs() (out, in uint64) {
+	w.tipMtx.RLock()
+	tip := w.currentTip.Number.Uint64()
+	w.tipMtx.RUnlock()
+
+	isETH := w.assetID == BipID
+
+	addPendingTx := func(pt *pendingTx) {
+		in += pt.in
+		if isETH {
+			if pt.assetID == BipID {
+				out += pt.out + pt.maxFees
+			} else {
+				out += pt.maxFees
+			}
+			return
+		}
+		// token
+		out += pt.out
+	}
+
+	w.pendingTxMtx.Lock()
+	defer w.pendingTxMtx.Unlock()
+	for txHash, pt := range w.pendingTxs {
+		if !isETH && pt.assetID != w.assetID {
+			continue
+		}
+		if pt.lastCheck == tip {
+			// Expect nothing has changed since our last check.
+			addPendingTx(pt)
+			continue
+		}
+		confs, err := w.node.transactionConfirmations(w.ctx, txHash)
+		if err != nil {
+			if !errors.Is(err, asset.CoinNotFoundError) {
+				w.log.Errorf("Error getting confirmations for pending tx %s: %v", txHash, err)
+			}
+			if time.Since(pt.stamp) > time.Minute*5 {
+				currentNonce, err := w.node.getConfirmedNonce(w.ctx)
+				if err != nil {
+					w.log.Errorf("Error getting account nonce for stale pending tx check: %v", err)
+					continue
+				}
+				if currentNonce >= pt.nonce {
+					w.log.Errorf("pending tx not confirmed but nonce has been confirmed")
+					delete(w.pendingTxs, txHash)
+				}
+			}
+			if !errors.Is(err, asset.CoinNotFoundError) {
+				continue
+			}
+		}
+		if confs > 0 {
+			delete(w.pendingTxs, txHash)
+			continue
+		}
+		pt.lastCheck = tip // Avoid multiple checks on the same block.
+		addPendingTx(pt)
+	}
+	return
+}
+
 func (w *assetWallet) balanceWithTxPool() (*Balance, error) {
 	isETH := w.assetID == BipID
-	var confirmed *big.Int
-	var err error
-	if isETH {
-		confirmed, err = w.node.addressBalance(w.ctx, w.addr)
-	} else {
-		confirmed, err = w.tokenBalance()
+
+	getBal := func() (*big.Int, error) {
+		if isETH {
+			return w.node.addressBalance(w.ctx, w.addr)
+		} else {
+			return w.tokenBalance()
+		}
 	}
+
+	txPoolNode, is := w.node.(txPoolFetcher)
+	if !is {
+		var confirmed *big.Int
+		for {
+			if confirmed == nil {
+				var err error
+				confirmed, err = getBal()
+				if err != nil {
+					return nil, fmt.Errorf("balance error: %v", err)
+				}
+			}
+			out, in := w.sumPendingTxs()
+			checkBal, err := getBal()
+			if err != nil {
+				return nil, fmt.Errorf("balance consistency check error: %v", err)
+			}
+			outEVM := w.evmify(out)
+			// If our balance has gone down in the interim, we'll use the lower
+			// balance, but ensure that we're not setting up an underflow for
+			// available balance.
+			if checkBal.Cmp(confirmed) != 0 {
+				w.log.Debugf("balance changed while checking pending txs. Trying again.")
+				confirmed = checkBal
+				continue
+			}
+			if outEVM.Cmp(confirmed) > 0 {
```

Comment:
Why not just return a 0 available balance?

Reply:
We don't set available balance here. It's calculated in `(*assetWallet).balance`. We could add an underflow check there for good measure, but I still think it's appropriate to check consistency here. if the `checkBal` matches up and the `outEVM` is still too high, that would indicate to me that we have a serious shortcoming in our logic, which is error-worthy, imo.